platform/pulpcore/app/models/storage.py

```diff
@@ -225,176 +136,35 @@ def _save(path, content):
 
         return path
 
-    @staticmethod
-    def get_artifact_path(sha256digest):
-        """
-        Determine the absolute path where a file backing the Artifact should be stored.
-
-        Args:
-            sha256digest (str): sha256 digest of the file for the Artifact
-
-        Returns:
-            A string representing the absolute path where a file backing the Artifact should be
-            stored
-        """
-        return os.path.join(settings.MEDIA_ROOT, 'artifact', sha256digest[0:2], sha256digest[2:])
-
-    @staticmethod
-    def published_metadata_path(model, name):
-        """
-        Get the storage path for published metadata.
-
-        Args:
-            model (pulpcore.app.models.PublishedMetadata): A model instance.
-            name (str): The file name.
-
-        Returns:
-            str: The absolute storage path.
-        """
-        return os.path.join(
-            settings.MEDIA_ROOT,
-            'published',
-            'metadata',
-            str(model.pk),
-            name)
-
-    def get_available_name(self, name, max_length=None):
-        """
-        Get the available absolute path based on the name requested.
-        Required by the Storage API.
-
-        Args:
-            name (str): File name.
-            max_length (int): Maximum length, in characters, of the returned path.
-
-        Returns:
-            str: Available name.
-
-        Raises:
-            ValueError if ``max_length`` exceeded.
-        """
-        if max_length:
-            if len(name) > max_length:
-                raise ValueError('max_length exceeded')
-        return name
-
-    def delete(self, path):
-        """
-        Delete the file at the specified path.
-        Required by the Storage API.
 
-        Args:
-            path (str): An absolute file path.
-        """
-        FileSystem.unlink(path)
-
-    def exists(self, path):
-        """
-        Get whether the file actually exists at the specified path.
-        Required by the Storage API.
-
-        Args:
-            path (str): An absolute file path.
-
-        Returns:
-            bool: True if exists.
-        """
-        return os.path.exists(path)
-
-    def listdir(self, path):
-        """
-        List the content of the directory at the specified path.
-        Required by the Storage API.
-
-        Args:
-            path (str): An absolute directory path.
-
-        Returns:
-            A tuple of two lists: (directories, files).
-        """
-        files = []
-        directories = []
-        for entry in os.listdir(path):
-            if os.path.isdir(os.path.join(path, entry)):
-                directories.append(entry)
-            else:
-                files.append(entry)
-        return (directories, files)
-
-    def path(self, name):
```

Comment:
The `FileSystemStorage.path` uses a `safe_path()` that does 2 things.

1. Ensure that absolute paths are within the MEDIA_ROOT.  This is good.

2. For relative paths, it joins the MEDIA_ROOT & path.  This is bad.  For artifacts, we never want the storage to change the storage location.

This should be overriden to raise an exception whenever called.

Reply:
I don't think we need to worry about enforcing the paths. I don't think plugin writers will be providing paths for where to save files. I am not 100% sure on this, so if you can provide an example of when we need to guard against this, that would help me understand this requirement better.

Reply:
I don't think making assumptions about how the class is used is a good practice.  This storage class may be used by every model in pulp.  Both core and plugin models.  Plugin writers could add `FileField` to any of their models.  Passing a relative path could be unintentional by the plugin writer or a coding error.

The `path()` method goes beyond the `Storage` API and adds the behavior of joining relative paths to the MEDIA_ROOT.  There is no guarantee that any other storage back-end will do this so pulp should not rely on it.  Looking at S3 storage back-end, it uses the path *as-is* for the object key.

We should continue to override this method as suggested (or as it originally was) to ensure consistent, deterministic  behavior across storage back-ends.